### PR TITLE
CI: Pre-commit to apply black/isort/flake8 on ipython notebooks

### PR DESCRIPTION
**Pull request by cphyc**
## PR Summary

This adds a pre-commit hook to automatically format with black our notebooks. This should respect our black/isort/pyupgrade configs.

### CODE REVIEW

**Comment by cphyc** [comment]
pre-commit.ci run

**Comment by neutrinoceros** [comment]
Depending on the side of the diff we'll get from this (likely, big), I'd advise that this would be postpone to after the incoming release.

**Comment by cphyc** [comment]
pre-commit autofix

**Comment by cphyc** [comment]
@neutrinoceros I triggered the pre-commit.ci bot, but actually we could decide _not to_ activate it and create an issue to have it triggered once 4.0 is released? Wdyt?

**Comment by neutrinoceros** [comment]
> We could create an issue to have it triggered once 4.0 is released? Wdyt?

I'd put it in the 4.1 Milestone, drop the auto-fix commit for now, and switch this to draft for now

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -59,3 +59,19 @@ repos:
...
+  hooks:
+    - id: nbqa-pyupgrade
+      args: [--nbqa-mutate, --py36-plus]
+    - id: nbqa-black
+      args: [--nbqa-mutate]
+      language_version: python3
+    - id: nbqa-isort
+      args: [--nbqa-mutate]
+    - id: nbqa-flake8
+      args: [--extend-ignore=E402]
```
Comment (cphyc) [inline .pre-commit-config.yaml:73]: Note: deactivating E402 (https://www.flake8rules.com/rules/E402.html) as it seems to clash with adding some line magics in cells.

**Comment by cphyc** [comment]
Todo:
- [ ] once we're happy with this PR, call `pre-commit.ci autofix`
- [ ] inspect whether we can reactivate E402 with flake or not

**Comment by MarcoGorelli** [comment]
Hey, thanks for trying out nbQA!

Just a note on one of the error messages:

> doc/source/analyzing/mesh_filter.ipynb:cell_4:4:5: E999 SyntaxError: invalid syntax

This looks like a bug in nbQA - cell4, after `black` has been run, becomes
```python
print("Temperature of all cells:\n ad['temperature'] = \n%s\n" % ad["temperature"])
print(
    "Temperatures of all \"hot\" cells:\n hot_ad['temperature'] = \n%s"
    % hot_ad["temperature"]
)
```
and that's definitely valid syntax...I'll come back to you on this, sorry about that, but thanks for helping to find this bug!

**Comment by MarcoGorelli** [comment]
Got it, it's because `IPython.core.inputsplitter.IPythonInputSplitter` thinks "`% hot_ad["temperature"]`" is IPython magic:
```python
(Pdb) !line
'    % hot_ad["temperature"]\n'
(Pdb) IPythonInputSplitter().transform_cell(line)
'get_ipython().run_line_magic(\'hot_ad\', \'["temperature"]\')\n'
```

I think the solution is to stop trying to handle inline magics, it's just too complicated / unreliable and leads to bugs like this one. I'll ping when a new version's out, hopefully this weekend!

**Comment by neutrinoceros** [comment]
Thanks a lot @MarcoGorelli !
There is no rush on our side, FYI :)

**Comment by MarcoGorelli** [comment]
> Thanks a lot @MarcoGorelli !
> There is no rush on our side, FYI :)

Fixed in version 0.9.0 :tada: 

However, it looks like `doc/source/cookbook/fits_xray_images.ipynb` now has a syntax error:
```python
error: cannot format doc/source/cookbook/fits_xray_images.ipynb: Cannot parse: cell_12:2:125:                    [("fits", "flux"), ("fits", "projected_temperature"), ("gas", "pseudo_pressure"), ("gas", "pseudo_entropy"]), 
Oh no! 💥 💔 💥
```

Looks like there's a missing bracket after "pseudo_entropy"?

**Comment by neutrinoceros** [comment]
Awesome, thank you so much for catching this !

**Comment by MarcoGorelli** [comment]
Happy to help, and thanks for your awesome library!

`doc/source/cookbook/fits_xray_images.ipynb` is still failing as it contains invalid syntax - I'd suggest either fixing that first, or (for now) using
```diff
-      args: [--nbqa-mutate]
+     args: [--nbqa-mutate, --nbqa-skip-bad-cells]
```

**Comment by cphyc** [comment]
@MarcoGorelli thanks! I've fixed the syntax error and the flake8 issues, it now seems to be working like a charm!

**Comment by cphyc** [comment]
pre-commit.ci autofix

**Comment by neutrinoceros** [comment]
@cphyc I believe the failures on Jenkins were fixed on the main branch, can you rebase this please ?

**Comment by cphyc** [comment]
pre-commit.ci autofix

**Comment by cphyc** [comment]
pre-commit.ci autofix

**Comment by cphyc** [comment]
For reviewers: you may want to only see the changes I made and not the one pre-commit did (https://github.com/yt-project/yt/pull/3287/files/a5cb16a652ec8ae35f4a6eb907d1133dbb9a7cde)

**Review by neutrinoceros — APPROVED** [review]
LGTM

**Comment by cphyc** [comment]
pre-commit.ci autofix

**Comment by cphyc** [comment]
pre-commit.ci autofix

**Comment by neutrinoceros** [comment]
pre-commit.ci autofix

**Comment by neutrinoceros** [comment]
@cphyc looks like pre-commit.ci auto fixing is currently broken

**Comment by neutrinoceros** [comment]
Nevermind I get what's happening: most likely pre-commit.ci is timing out after 3min so it can't apply your changes.
If you don't want to assume authorship for a this sweeping change, I suggest you use a similar trick as I did with #2749 :
```shell
git commit --author=convert-repo ...
```

**Comment by neutrinoceros** [comment]
Wow, black 21.8b0 now has a specialised pre-commit hook for notebooks,
https://github.com/psf/black/pull/2357

@MarcoGorelli since you contributed this too, what's your recommendation ? is there any edge to keep using the nbqa hook  or should we switch ?

**Comment by MarcoGorelli** [comment]
Hey @neutrinoceros, thanks for asking - I'd advise using black's hook, I'll deprecate the `nbqa-black` one

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -37,6 +37,9 @@ repos:
     hooks:
     - id: black
       language_version: python3
+    - id: black-jupyter
+      language_version: python3
```
Comment (MarcoGorelli) [inline .pre-commit-config.yaml:41]: `black-jupyter` does everything `black` does, but also checks notebooks - you probably just need that one

Comment (neutrinoceros) [reply .pre-commit-config.yaml]: While you're at it @cphyc, you can remove the `language_version` line. It became useless in the latest release.

**Pull request merged**